Read the following data set:<br>
https://archive.ics.uci.edu/ml/machine-learning-databases/adult/<br>
Rename the columns as per the description from this file:<br>
https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.names

In [16]:
import numpy as np 
import pandas as pd
from pandasql import sqldf

Read the following data set:<br>
https://archive.ics.uci.edu/ml/machine-learning-databases/adult/<br>
Rename the columns as per the description from this file:<br>
https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.names

In [127]:
url="https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data"
df=pd.read_csv(url, header=None,
               names=['age','workclass','fnlwgt','education','education-num','marital-status','occupation','relationship',
                        'race','sex', 'capital-gain','capital-loss','hours-per-week', 'native-country','class'])


Create a sql db from adult dataset and name it sqladb

In [128]:
sqladb=df

1. Select 10 records from the adult sqladb

In [129]:
TenRec=sqldf("select * from sqladb LIMIt 10")
print(TenRec, end='')

   age          workclass  fnlwgt   education  education-num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   
5   37            Private  284582     Masters             14   
6   49            Private  160187         9th              5   
7   52   Self-emp-not-inc  209642     HS-grad              9   
8   31            Private   45781     Masters             14   
9   42            Private  159449   Bachelors             13   

           marital-status          occupation    relationship    race  \
0           Never-married        Adm-clerical   Not-in-family   White   
1      Married-civ-spouse     Exec-managerial         Husband   White   
2                Divorced   Handlers-cleaners   Not-in-family   White   
3  

2. Show me the average hours per week of all men who are working in private sector

In [130]:
sql2="select workclass,avg([hours-per-week]) from sqladb where workclass like '%Private' group by workclass"
avghr=sqldf(sql2)
avghr

workclass       avg(
0   Private  40.267096

3. Show me the frequency table for education, occupation and relationship, separately

$Frequency Table For Education:$

In [131]:
sql3a="select education, count(education) from sqladb group by education"
edu_freq=sqldf(sql3a)
edu_freq

education  count(education)
0            10th               933
1            11th              1175
2            12th               433
3         1st-4th               168
4         5th-6th               333
5         7th-8th               646
6             9th               514
7      Assoc-acdm              1067
8       Assoc-voc              1382
9       Bachelors              5355
10      Doctorate               413
11        HS-grad             10501
12        Masters              1723
13      Preschool                51
14    Prof-school               576
15   Some-college              7291

$FrequencyTableForOccupation:$

In [132]:
sql3b="select occupation, count(occupation) from sqladb group by occupation"
occ_freq=sqldf(sql3b)
occ_freq

occupation  count(occupation)
0                    ?               1843
1         Adm-clerical               3770
2         Armed-Forces                  9
3         Craft-repair               4099
4      Exec-managerial               4066
5      Farming-fishing                994
6    Handlers-cleaners               1370
7    Machine-op-inspct               2002
8        Other-service               3295
9      Priv-house-serv                149
10      Prof-specialty               4140
11     Protective-serv                649
12               Sales               3650
13        Tech-support                928
14    Transport-moving               1597

$FrequencyTableForRelationship:$

In [133]:
sql3c="select relationship, count(relationship) from sqladb group by relationship"
rel_freq=sqldf(sql3c)
rel_freq

relationship  count(relationship)
0          Husband                13193
1    Not-in-family                 8305
2   Other-relative                  981
3        Own-child                 5068
4        Unmarried                 3446
5             Wife                 1568

4. Are there any people who are married, working in private sector and having a masters
degree

In [134]:
#Check for different marital-status available
sql4a="select distinct [marital-status] from sqladb "
sqldf(sql4a)

marital-status
0           Never-married
1      Married-civ-spouse
2                Divorced
3   Married-spouse-absent
4               Separated
5       Married-AF-spouse
6                 Widowed

In [136]:
#ignore marital-status in Never-married,Divorced,Separated,Widowed
#check for relationship status after ignoring above marital-status
sql4b="select distinct [marital-status],relationship from sqladb where [marital-status] not in (' Never-married',' Divorced',' Separated',' Widowed')"
sqldf(sql4b)

marital-status     relationship
0       Married-civ-spouse          Husband
1       Married-civ-spouse             Wife
2    Married-spouse-absent    Not-in-family
3       Married-civ-spouse        Own-child
4        Married-AF-spouse             Wife
5       Married-civ-spouse   Other-relative
6    Married-spouse-absent        Unmarried
7    Married-spouse-absent        Own-child
8    Married-spouse-absent   Other-relative
9       Married-civ-spouse    Not-in-family
10       Married-AF-spouse          Husband
11       Married-AF-spouse        Own-child
12       Married-AF-spouse   Other-relative

In [137]:
#ignore marital-status in Never-married,Divorced,Separated,Widowed
#check for relationship status after ignoring above marital-status
sql4b="select distinct [marital-status],relationship from sqladb where [marital-status] not in (' Never-married',' Divorced',' Separated',' Widowed')"
sqldf(sql4b)

marital-status     relationship
0       Married-civ-spouse          Husband
1       Married-civ-spouse             Wife
2    Married-spouse-absent    Not-in-family
3       Married-civ-spouse        Own-child
4        Married-AF-spouse             Wife
5       Married-civ-spouse   Other-relative
6    Married-spouse-absent        Unmarried
7    Married-spouse-absent        Own-child
8    Married-spouse-absent   Other-relative
9       Married-civ-spouse    Not-in-family
10       Married-AF-spouse          Husband
11       Married-AF-spouse        Own-child
12       Married-AF-spouse   Other-relative

In [138]:
#ignore  relationship=Unmarried
sql4c="select distinct [marital-status],relationship from sqladb where [marital-status] not in (' Never-married',' Divorced',' Separated',' Widowed') and relationship<>' Unmarried'"
sqldf(sql4c)

marital-status     relationship
0       Married-civ-spouse          Husband
1       Married-civ-spouse             Wife
2    Married-spouse-absent    Not-in-family
3       Married-civ-spouse        Own-child
4        Married-AF-spouse             Wife
5       Married-civ-spouse   Other-relative
6    Married-spouse-absent        Own-child
7    Married-spouse-absent   Other-relative
8       Married-civ-spouse    Not-in-family
9        Married-AF-spouse          Husband
10       Married-AF-spouse        Own-child
11       Married-AF-spouse   Other-relative

In [143]:
#find count of all records satisfying given criteria
sql4d="select count(*) from sqladb where [marital-status] not in (' Never-married',' Divorced',' Separated',' Widowed') and relationship<>' Unmarried' and education=' Masters' and workclass=' Private'"
sqldf(sql4d)



count(*)
0       539

$CONCLUSION:$

There are 539 records where the people who are married, working in private sector and having a masters degree

5. What is the average, minimum and maximum age group for people working in
different sectors

In [144]:
sql5="select occupation,avg(age),max(age),min(age) from sqladb group by occupation"
sqldf(sql5)

occupation   avg(age)  max(age)  min(age)
0                    ?  40.882800        90        17
1         Adm-clerical  36.964456        90        17
2         Armed-Forces  30.222222        46        23
3         Craft-repair  39.031471        90        17
4      Exec-managerial  42.169208        90        17
5      Farming-fishing  41.211268        90        17
6    Handlers-cleaners  32.165693        90        17
7    Machine-op-inspct  37.715285        90        17
8        Other-service  34.949621        90        17
9      Priv-house-serv  41.724832        81        17
10      Prof-specialty  40.517633        90        17
11     Protective-serv  38.953775        90        17
12               Sales  37.353973        90        17
13        Tech-support  37.022629        73        17
14    Transport-moving  40.197871        90        17

6. Calculate age distribution by country

In [154]:
sql6="select [native-country],Age_interval,count(*) from(select [native-country],age,case when age<=25 then '1-25'when age between 26  and 50 then '26-50' else 'above 50' end as'Age_interval' from sqladb) group by Age_interval,[native-country]"
sqldf(sql6)

native-country Age_interval  count(*)
0                              ?         1-25        79
1                       Cambodia         1-25         2
2                         Canada         1-25        18
3                          China         1-25         6
4                       Columbia         1-25        10
5                           Cuba         1-25         6
6             Dominican-Republic         1-25        16
7                        Ecuador         1-25         5
8                    El-Salvador         1-25        36
9                        England         1-25        12
10                        France         1-25         2
11                       Germany         1-25        16
12                        Greece         1-25         4
13                     Guatemala         1-25        22
14                         Haiti         1-25         9
15                      Honduras         1-25         2
16                          Hong         1-25         1
17                       Hungary         1-25         1
18                         India         1-25        12
19                          Iran         1-25         4
20                       Ireland         1-25         2
21                         Italy         1-25         7
22                       Jamaica         1-25        18
23                         Japan         1-25         4
24                          Laos         1-25         4
25                        Mexico         1-25       188
26                     Nicaragua         1-25         8
27    Outlying-US(Guam-USVI-etc)         1-25         3
28                          Peru         1-25         9
29                   Philippines         1-25        26
..                           ...          ...       ...
94                       Germany     above 50        28
95                        Greece     above 50        14
96                     Guatemala     above 50         6
97                         Haiti     above 50        11
98                      Honduras     above 50         1
99                          Hong     above 50         1
100                      Hungary     above 50         7
101                        India     above 50        17
102                         Iran     above 50         7
103                      Ireland     above 50         3
104                        Italy     above 50        30
105                      Jamaica     above 50        12
106                        Japan     above 50         6
107                         Laos     above 50         2
108                       Mexico     above 50        61
109                    Nicaragua     above 50         3
110   Outlying-US(Guam-USVI-etc)     above 50         3
111                         Peru     above 50         2
112                  Philippines     above 50        33
113                       Poland     above 50        17
114                     Portugal     above 50         8
115                  Puerto-Rico     above 50        23
116                     Scotland     above 50         3
117                        South     above 50        15
118                       Taiwan     above 50         4
119                     Thailand     above 50         1
120              Trinadad&Tobago     above 50         3
121                United-States     above 50      5866
122                      Vietnam     above 50         8
123                   Yugoslavia     above 50         3

[124 rows x 3 columns]

7. Compute a new column as 'Net-Capital-Gain' from the two columns 'capital-gain' and
'capital-loss'

In [156]:
sql7="select 'capital-gain' - 'capital-loss' as 'Net-Capital-Gain' from sqladb"
sqldf(sql7)

Net-Capital-Gain
0                     0
1                     0
2                     0
3                     0
4                     0
5                     0
6                     0
7                     0
8                     0
9                     0
10                    0
11                    0
12                    0
13                    0
14                    0
15                    0
16                    0
17                    0
18                    0
19                    0
20                    0
21                    0
22                    0
23                    0
24                    0
25                    0
26                    0
27                    0
28                    0
29                    0
...                 ...
32531                 0
32532                 0
32533                 0
32534                 0
32535                 0
32536                 0
32537                 0
32538                 0
32539                 0
32540                 0
32541                 0
32542                 0
32543                 0
32544                 0
32545                 0
32546                 0
32547                 0
32548                 0
32549                 0
32550                 0
32551                 0
32552                 0
32553                 0
32554                 0
32555                 0
32556                 0
32557                 0
32558                 0
32559                 0
32560                 0

[32561 rows x 1 columns]